pkg/controller/networkpolicy/endpoint_querier.go

```diff
@@ -155,50 +189,159 @@ func (eq *endpointQuerier) QueryNetworkPolicies(namespace string, podName string
-			},
-			Direction: cpv1beta.DirectionOut,
-			RuleIndex: internalPolicy.index,
-		}
-		responseRules = append(responseRules, newRule)
-	}
-	for _, internalPolicy := range ingress {
-		newRule := Rule{
-			PolicyRef: PolicyRef{
-				Namespace: internalPolicy.policy.SourceRef.Namespace,
-				Name:      internalPolicy.policy.SourceRef.Name,
-				UID:       internalPolicy.policy.SourceRef.UID,
-			},
-			Direction: cpv1beta.DirectionIn,
-			RuleIndex: internalPolicy.index,
-		}
-		responseRules = append(responseRules, newRule)
-	}
-	// for now, selector only selects a single endpoint (pod, namespace)
-	endpoint := Endpoint{
-		Namespace: namespace,
-		Name:      podName,
-		Policies:  responsePolicies,
-		Rules:     responseRules,
-	}
-	return &EndpointQueryResponse{[]Endpoint{endpoint}}, nil
+	return &EndpointNetworkPolicyRules{namespace, podName, applied, ingress, egress}, nil
+}
+
+// processQueryResults processes the AppliedPolicies in endpoints, returns a set
+// of the policy UIDs, and manually inserts Kubernetes NetworkPolicy default isolation
+// rules if exists. The default isolation rule's direction depends on ifSource,
+// and has the index of -1 to indicate lower precedence.
+func processQueryResults(endpoints *EndpointNetworkPolicyRules, ifSource bool) (sets.Set[types.UID], []*RuleInfo) {
+	policyUIDs := sets.New[types.UID]()
+	isolation := make([]*RuleInfo, 0)
+	for _, internalPolicy := range endpoints.AppliedPolicies {
+		policyUIDs.Insert(internalPolicy.SourceRef.UID)
+		if internalPolicy.SourceRef.Type == controlplane.K8sNetworkPolicy {
+			// check if the Kubernetes NetworkPolicy creates ingress or egress isolation
+			for _, rule := range internalPolicy.Rules {
+				if rule.Direction == controlplane.DirectionIn && !ifSource {
+					isolation = append(isolation, &RuleInfo{Policy: internalPolicy, Index: -1,
+						Rule: &controlplane.NetworkPolicyRule{Direction: rule.Direction, Name: rule.Name, Action: rule.Action}})
+				} else if rule.Direction == controlplane.DirectionOut && ifSource {
+					isolation = append(isolation, &RuleInfo{Policy: internalPolicy, Index: -1,
+						Rule: &controlplane.NetworkPolicyRule{Direction: rule.Direction, Name: rule.Name, Action: rule.Action}})
+				}
+			}
+		}
+	}
+	return policyUIDs, isolation
+}
+
+// predictEndpointsRules returns the predicted rules effective from srcEndpoints to dstEndpoints.
+// Rules returned satisfy a. in source applied policies and destination egress rules,
+// or b. in source ingress rules and destination applied policies or c. applied to KNP default isolation.
+func predictEndpointsRules(srcEndpoints, dstEndpoints *EndpointNetworkPolicyRules) (commonRule *RuleInfo) {
+	commonRules := make([]*RuleInfo, 0)
+	if srcEndpoints != nil && dstEndpoints != nil {
+		srcPolicies, srcIsolated := processQueryResults(srcEndpoints, true)
+		dstPolicies, dstIsolated := processQueryResults(dstEndpoints, false)
+		for _, rule := range dstEndpoints.EgressRules {
```

Comment:
The `dstEndpoints.EgressRules` naming got me really confused when reviewing. After jumping around a bunch of functions I realized that the rules stored there are "egress rules where the endpoint is selected as the destination of the rule", which is drastically different then what the field name suggests. I would change it to something like `EndpointAsEgressDstRules`, and similarly, `EndpointAsIngressSrcRules`, which would convey the point right away

Reply:
Renamed.